datafusion/expr/src/utils.rs

```diff
@@ -45,6 +45,22 @@ pub fn exprlist_to_columns(expr: &[Expr], accum: &mut HashSet<Column>) -> Result
     Ok(())
 }
 
+/// Count the number of distinct exprs in a list of group by expressions. If the
+/// first element is a `GroupingSet` expression then it must be the only expr.
+pub fn grouping_set_expr_count(group_expr: &[Expr]) -> Result<usize> {
+    if let Some(Expr::GroupingSet(grouping_set)) = group_expr.first() {
```

Comment:
It's weird that this function takes either a GroupingSet expression, or an arbitrary expression. I dealt with this a lot in the subquery decorrelation - it almost feels like dynamic typing. I think the root cause might be that many of the enum values have fields inside them, rather than a struct - so it's impossible to de-reference them and pass them between functions.

How would you feel about blanket converting all enum values like `InSubquery`, `Exists`, etc to point to structs?

Reply:
I also found this frustrating to work with. Maybe we could introduce another enum?

```
pub enum Grouping {
    Expr(Vec<Expr>),
    Set(GroupingSet)
}
```

and then

```
pub struct Aggregate {
    pub input: Arc<LogicalPlan>,
    pub group_expr: Grouping,
    pub aggr_expr: Vec<Expr>,
    pub schema: DFSchemaRef,
}
```

Reply:
> How would you feel about blanket converting all enum values like `InSubquery`, `Exists`, etc to point to structs?

Sounds good to me. This would be consistent with how we do things in the `LogicalPlan`.